### PR TITLE
Add heyoka.py

### PR DESCRIPTION
**Pull request by bluescarni**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by bluescarni** [comment]
@isuruf would you mind taking a look?

File: recipes/heyoka.py/build.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+
+mkdir build
+cd build
+
+if [[ "$(uname)" == "Darwin" ]]; then
```
Comment (isuruf) [inline recipes/heyoka.py/build.sh:6]:
```suggestion
if [[ "$target_platform" == osx-* ]]; then
```

File: recipes/heyoka.py/build.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    export AR_CMAKE_SETTING=
+    export RANLIB_CMAKE_SETTING=
+else
+    # Workaround for making the LTO machinery work on Linux.
```
Comment (isuruf) [inline recipes/heyoka.py/build.sh:10]: No LTO on Linux?

File: recipes/heyoka.py/build.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    export AR_CMAKE_SETTING="-DCMAKE_CXX_COMPILER_AR=$GCC_AR -DCMAKE_C_COMPILER_AR=$GCC_AR"
+    export RANLIB_CMAKE_SETTING="-DCMAKE_CXX_COMPILER_RANLIB=$GCC_RANLIB -DCMAKE_C_COMPILER_RANLIB=$GCC_RANLIB"
+fi
+
+cmake \
```
Comment (isuruf) [inline recipes/heyoka.py/build.sh:15]:
```suggestion
cmake ${CMAKE_ARGS}\
```

File: recipes/heyoka.py/build.sh (hunk 2 above)
Comment (bluescarni) [reply recipes/heyoka.py/build.sh:10]: Do you know which are the correct settings for these variables under OSX? The build system is trying to enable LTO on OSX as well, but failing to do so:

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=277482&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=1162

**Comment by bluescarni** [comment]
@isuruf Thanks a lot for the quick review!

File: recipes/heyoka.py/build.sh (hunk 2 above)
Comment (bluescarni) [reply recipes/heyoka.py/build.sh:10]: Ahh now that I am reading the output of the azure builds more carefully, is ``${CMAKE_ARGS}`` supposed to take care of seeing AR, RANLIB, etc. as well? In such case, perhaps I can remove the hack of setting them manually in the Linux builds?

Comment (isuruf) [reply recipes/heyoka.py/build.sh:10]: No, they use just the generic AR and RANLIB, so you should keep this.

Is there a more detailed error message about OSX?

Comment (bluescarni) [reply recipes/heyoka.py/build.sh:10]: I don't have access to an OSX machine, I would not know how to produce a more detailed error message.

I have tried setting explicitly AR/RANLIB in the same fashion as done in linux, let's see if anything changes...

Comment (bluescarni) [reply recipes/heyoka.py/build.sh:10]: @isuruf the latest commit seems to have done it, but it does not seem like a very solid solution as the AR/RANLIB versions are hard-coded in the script...

File: recipes/heyoka.py/build.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+
+mkdir build
+cd build
+
+if [[ "$target_platform" == osx-* ]]; then
+    export AR_CMAKE_SETTING="-DCMAKE_CXX_COMPILER_AR=$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-ar -DCMAKE_C_COMPILER_AR=$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-ar"
+    export RANLIB_CMAKE_SETTING="-DCMAKE_CXX_COMPILER_RANLIB=$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-ranlib -DCMAKE_C_COMPILER_RANLIB=$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-ranlib"
```
Comment (isuruf) [inline recipes/heyoka.py/build.sh:8]:
```suggestion
    export AR_CMAKE_SETTING="-DCMAKE_CXX_COMPILER_AR=$AR"
    export RANLIB_CMAKE_SETTING="-DCMAKE_CXX_COMPILER_RANLIB=$RANLIB -DCMAKE_C_COMPILER_RANLIB=$RANLIB"
```

File: recipes/heyoka.py/build.sh (hunk 2 above)
Comment (isuruf) [reply recipes/heyoka.py/build.sh:10]: > I don't have access to an OSX machine, I would not know how to produce a more detailed error message.

You can cross compile if you'd like to.

File: recipes/heyoka.py/build.sh (hunk 4 above)
Comment (bluescarni) [reply recipes/heyoka.py/build.sh:8]: @isuruf I think I tried this at 

https://github.com/conda-forge/staged-recipes/pull/13961/commits/c161caa26904391a894aff8d569b413ede5e8b47

but it did not seem to work?

Comment (isuruf) [reply recipes/heyoka.py/build.sh:8]: Can you send a PR to add these to https://github.com/conda-forge/clang-compiler-activation-feedstock/blob/master/recipe/activate-clang.sh#L131? I think that might be easier

Comment (bluescarni) [reply recipes/heyoka.py/build.sh:8]: Gonna try again...

**Comment by bluescarni** [comment]
@isuruf using ``$AR/$RANLIB`` does not seem to work, CMake errors out saying:

```console
-- IPO requested, but it is not supported by the compiler:
Change Dir: $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin

Run Build Command(s):/usr/bin/make && $BUILD_PREFIX/bin/cmake -S$SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/src -B$SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin --check-build-system CMakeFiles/Makefile.cmake 0
$BUILD_PREFIX/bin/cmake -E cmake_progress_start $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin/CMakeFiles $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin//CMakeFiles/progress.marks
/Applications/Xcode_12.3.app/Contents/Developer/usr/bin/make  -f CMakeFiles/Makefile2 all
/Applications/Xcode_12.3.app/Contents/Developer/usr/bin/make  -f CMakeFiles/foo.dir/build.make CMakeFiles/foo.dir/depend
cd $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin && $BUILD_PREFIX/bin/cmake -E cmake_depends "Unix Makefiles" $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/src $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/src $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/bin/CMakeFiles/foo.dir/DependInfo.cmake
Scanning dependencies of target foo
/Applications/Xcode_12.3.app/Contents/Developer/usr/bin/make  -f CMakeFiles/foo.dir/build.make CMakeFiles/foo.dir/build
[ 25%] Building CXX object CMakeFiles/foo.dir/foo.cpp.o
$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++   -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/heyoka.py-0.3.0 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -flto=thin -isysroot /Applications/Xcode_12.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -mmacosx-version-min=10.9 -o CMakeFiles/foo.dir/foo.cpp.o -c $SRC_DIR/build/heyoka/CMakeFiles/_CMakeLTOTest-CXX/src/foo.cpp
[ 50%] Linking CXX static library libfoo.a
$BUILD_PREFIX/bin/cmake -P CMakeFiles/foo.dir/cmake_clean_target.cmake
$BUILD_PREFIX/bin/cmake -E cmake_link_script CMakeFiles/foo.dir/link.txt --verbose=1
"$SRC_DIR/build/x86_64-apple-darwin13.4.0-ar" cr libfoo.a CMakeFiles/foo.dir/foo.cpp.o
Error running link command: No such file or directory
make[2]: *** [libfoo.a] Error 2
make[1]: *** [CMakeFiles/foo.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by isuruf** [comment]
Can you send a PR to add these to https://github.com/conda-forge/clang-compiler-activation-feedstock/blob/master/recipe/activate-clang.sh#L131?

**Comment by bluescarni** [comment]
@isuruf 

https://github.com/conda-forge/clang-compiler-activation-feedstock/pull/54

**Comment by isuruf** [comment]
Thanks @bluescarni. Can you send a PR to https://github.com/conda-forge/ctng-compiler-activation-feedstock/blob/master/recipe/activate-gcc.sh#L137 for gcc as well?

**Comment by bluescarni** [comment]
> Thanks @bluescarni. Can you send a PR to https://github.com/conda-forge/ctng-compiler-activation-feedstock/blob/master/recipe/activate-gcc.sh#L137 for gcc as well?

https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/40

**Comment by bluescarni** [comment]
@isuruf it seems like now LTO is working on OSX without any special setting in ``build.sh``. Shall I wait for the GCC PR to be merged so that I can remove also the LTO workaround in the linux build?

**Comment by bluescarni** [comment]
@isuruf do you think this is ready to go?

**Comment by bluescarni** [comment]
@isuruf thanks a lot!

**Pull request merged**